agent/taskresource/ssmsecret/ssmsecret.go

```diff
@@ -0,0 +1,476 @@
+	seelog.Infof("ssm secret resource: retrieving secrets for region %s in task: [%s]", region, secret.taskARN)
+	defer wg.Done()
+
+	var wgPerRegion sync.WaitGroup
+	var secretNames []string
+
+	for _, s := range secrets {
+		secretName := secret.extractNameFromValueFrom(s)
+		secretKey := secretName + "_" + region
+		if _, ok := secret.GetSSMSecretValue(secretKey); ok {
+			continue
+		}
+		secretNames = append(secretNames, secretName)
+		if len(secretNames) == 10 {
+			secretNamesTmp := make([]string, 10)
+			copy(secretNames, secretNamesTmp)
+			wgPerRegion.Add(1)
+			go secret.retrieveSSMSecretValues(region, secretNamesTmp, iamCredentials, &wgPerRegion, errorEvents)
+			secretNames = []string{}
+		}
+	}
+
+	if len(secretNames) > 0 {
+		wgPerRegion.Add(1)
+		go secret.retrieveSSMSecretValues(region, secretNames, iamCredentials, &wgPerRegion, errorEvents)
+	}
+
+	wgPerRegion.Wait()
+	return
+}
+
+func (secret *SSMSecretResource) retrieveSSMSecretValues(region string, names []string, iamCredentials credentials.IAMRoleCredentials, wg *sync.WaitGroup, errorEvents chan error) {
+	defer wg.Done()
+
+	ssmClient := secret.ssmClientCreator.NewSSMClient(region, iamCredentials)
+	seelog.Infof("ssm secret resource: retrieving resource for secrets %v in region [%s] in task: [%s]", names, region, secret.taskARN)
+	secValueMap, err := ssm.GetSecretsFromSSM(names, ssmClient)
+	if err != nil {
+		errorEvents <- err
+		return
+	}
+
+	secret.lock.Lock()
+	defer secret.lock.Unlock()
+
+	// put secret value in secretData
+	for secretName, secretValue := range secValueMap {
+		secretKey := secretName + "_" + region
+		secret.secretData[secretKey] = secretValue
+	}
+
+	return
+}
+
+func (secret *SSMSecretResource) extractNameFromValueFrom(secretData apicontainer.Secret) string {
+	valueFrom := secretData.ValueFrom
+	match, _ := regexp.MatchString(ParameterARNRegex, valueFrom)
+
+	if match {
+		index := strings.Index(valueFrom, ParameterARNPrefix)
+		return string(valueFrom[index+len(ParameterARNPrefix):])
+	} else {
+		return valueFrom
+	}
+
+}
+
+// GetRequiredSecrets returns the requiredSecrets field of ssmsecret task resource
+func (secret *SSMSecretResource) GetRequiredSecrets() map[string][]apicontainer.Secret {
+	secret.lock.RLock()
+	defer secret.lock.RUnlock()
+
+	return secret.requiredSecrets
+}
+
+// GetExecutionCredentialsID returns the execution role's credential ID
+func (secret *SSMSecretResource) GetExecutionCredentialsID() string {
+	secret.lock.RLock()
+	defer secret.lock.RUnlock()
+
+	return secret.executionCredentialsID
+}
+
+// Cleanup removes the secret value created for the task
+func (secret *SSMSecretResource) Cleanup() error {
+	secret.clearSSMSecretValue()
+	return nil
+}
+
+// clearSSMSecretValue cycles through the collection of secret value data and
+// removes them from the task
+func (secret *SSMSecretResource) clearSSMSecretValue() {
```

Comment:
Why do we need this? Won't this get garbage collected?

Reply:
Want to make sure, we remove all secret values.